lib/http1.c

```diff
@@ -662,32 +659,57 @@ static void handle_incoming_request(struct st_h2o_http1_conn_t *conn)
             if (h2o_http2_handle_upgrade(&conn->req, conn->super.connected_at) != 0)
                 h2o_send_error_400(&conn->req, "Invalid Request", "Broken upgrade request to HTTP/2", 0);
         } else if (entity_body_header_index != -1) {
-            /* Request has body, start reading it. Invocation of `h2o_process_request` is delayed to reduce backend concurrency. */
+            /* Request has body, start reading it */
+
             conn->req.timestamps.request_body_begin_at = h2o_gettimeofday(conn->super.ctx->loop);
-            if (expect.base != NULL) {
-                if (!h2o_lcstris(expect.base, expect.len, H2O_STRLIT("100-continue"))) {
+            if (create_entity_reader(conn, headers + entity_body_header_index) != 0)
+                return;
+            conn->_unconsumed_request_size = 0;
+            h2o_buffer_init(&conn->req_body, &h2o_socket_buffer_prototype);
+            conn->req.write_req.cb = write_req_first;
+            conn->req.write_req.ctx = &conn->req;
+
+            if (expect_header_index != -1) {
+                h2o_iovec_t expect_value = h2o_iovec_init(headers[expect_header_index].value, headers[expect_header_index].value_len);
+                if (!h2o_lcstris(expect_value.base, expect_value.len, H2O_STRLIT("100-continue"))) {
                     clear_timeouts(conn);
                     h2o_socket_read_stop(conn->sock);
                     h2o_send_error_417(&conn->req, "Expectation Failed", "unknown expectation",
                                        H2O_SEND_ERROR_HTTP1_CLOSE_CONNECTION);
                     return;
                 }
+                if (h2o_req_should_forward_expect(&conn->req)) {
+                    /* in forward mode, expect header is treated like other normal headers */
+                    char orig_case_cstr[sizeof("expect")] = {0};
+                    assert(sizeof(orig_case_cstr) - 1 == headers[expect_header_index].name_len);
+                    h2o_memcpy(orig_case_cstr, headers[expect_header_index].name, sizeof(orig_case_cstr) - 1);
+                    expect_value = h2o_strdup(&conn->req.pool, expect_value.base, expect_value.len);
+
+                    h2o_add_header(&conn->req.pool, &conn->req.headers, H2O_TOKEN_EXPECT, orig_case_cstr, expect_value.base, expect_value.len);
+
+                    h2o_buffer_consume(&conn->sock->input, reqlen);
+                    conn->req.write_req.cb = write_req_connect_first;
```

Comment:
We should come up with a better name, but nothing came to mind. Any suggestions?